edk2toolext/image_validation.py

```diff
@@ -0,0 +1,612 @@
+    pe.OPTIONAL_HEADER.DllCharacteristics = dllchar
+    return pe
+
+
+def get_nx_compat_flag(pe):
+    dllchar = pe.OPTIONAL_HEADER.DllCharacteristics
+
+    if has_characteristic(dllchar, 256):  # 256 (8th bit) is the mask
+        logging.info('True')
+        return 1
+    else:
+        logging.info('False')
+        return 0
+
+
+def clear_nx_compat_flag(pe):
+    dllchar = pe.OPTIONAL_HEADER.DllCharacteristics
+    dllchar = clear_bit(dllchar, 8)  # 8th bit is the nx_compat_flag
+    pe.OPTIONAL_HEADER.DllCharacteristics = dllchar
+    return pe
+
+
+def fill_missing_requirements(default, target):
+    for key in default:
+        if key not in target:
+            target[key] = default[key]
+    return target
+
+
+class Result:
+    PASS = '[PASS]'
+    WARN = '[WARNING]'
+    SKIP = '[SKIP]'
+    FAIL = '[FAIL]'
+
+
+class TestInterface:
+
+    def name(self):
+        """Returns the name of the test"""
+        raise NotImplementedError("Must Override Test Interface")
+
+    def execute(self, pe, config_data):
+        """
+        Executes the test
+
+        @param pe: The parser pefile
+        @param config_data: Configuration data for the specific target machine
+            and profile
+        """
+        raise NotImplementedError("Must Override Test Interface")
+
+
+class TestManager(object):
+    def __init__(self, config_data=None):
+        self.tests = []
+        if config_data:
+            self.config_data = config_data
+        else:
+            self.config_data = {
+                "TARGET_ARCH": {
+                    "X64": "IMAGE_FILE_MACHINE_AMD64",
+                    "IA32": "IMAGE_FILE_MACHINE_I386",
+                    "AARCH64": "IMAGE_FILE_MACHINE_ARM64",
+                    "ARM": "IMAGE_FILE_MACHINE_ARM"
+                },
+                "IMAGE_FILE_MACHINE_AMD64": {
+                    "DEFAULT": {
+                        "DATA_CODE_SEPARATION": True,
+                        "ALLOWED_SUBSYSTEMS": [
+                            "IMAGE_SUBSYSTEM_EFI_BOOT_SERVICE_DRIVER",
+                            "IMAGE_SUBSYSTEM_EFI_ROM"
+                        ],
+                        "ALIGNMENT": [
+                            {
+                                "COMPARISON": "==",
+                                "VALUE": 4096
+                            }
+                        ]
+                    },
+                    "APP": {
+                        "ALLOWED_SUBSYSTEMS": [
+                            "IMAGE_SUBSYSTEM_EFI_APPLICATION"
+                        ]
+                    },
+                    "DRIVER": {
+                        "ALLOWED_SUBSYSTEMS": [
+                            "IMAGE_SUBSYSTEM_EFI_BOOT_SERVICE_DRIVER",
+                            "IMAGE_SUBSYSTEM_EFI_RUNTIME_DRIVER"
+                            "IMAGE_SUBSYSTEM_EFI_ROM"
+                        ]
+                    },
+                    "PEI": {}
```

Comment:
Is "PEI" required if it is empty?

Reply:
Technically, no. However, I wanted to leave it in as the other machine types did have specific requirements for PEI. I'm fine with removing it if you think it is best.

Reply:
Some other sections show its usage and I think it's generally understood that it is not required if it is not used so I'd recommend removing it to reinforce that expectation.